Start producing bundles targeting ES2015 and ES5

wannabefro opened the pull request:
We want to remove our dependancy on polyfill.io for polyfilling our JavaScript. In order to accomplish this we want to once again start using `@babel/polyfills`. However we don't want to serve up more JavaScript than a browser needs (i.e. we don't need to polyfill `Promise` in latest Chrome)

In order to accomplish this, this PR makes it so that we are now producing a modern and legacy bundle. The modern bundle is targeting browsers that support `type="module"` and the legacy bundle browsers that don't.

We are taking advantage of `useBuiltIns: usage` so that we only include the polyfills that browsers need.

This PR is to support the main PR in the FL repo which will have the questions etc... This PR mainly gives us something to reference from it.

wannabefro committed:
Symlink src directory

We want to start producing multiple JS bundles so that we can optimise
what we send over the wire dependant on the browser (i.e. modern
browsers understand ES2015 so we shouldn't be sending the transpiled
ES5).

Thankfully Webpack makes this easy to accomplish.

Unfortunately that was before we decided to remove the need for packs
and move to a world where we dynamically import all our React
components.

Why does this make it harder?

There's no way (as far as I can tell) to tell Webpack to dynamically
import something that it has outputted, it can only dynamically import a
real file on disk (that it will then treat as an extra entry path and
run it through it's various pipelines).

So this means that we're left in a position where we can produce
multiple bundles but have no way of dynamically importing them :sadface:

How can we make this work?

If Webpack needs for there to be a real file on disk for it to
dynamically import, we can symlink `app/javascript` to
`app/javascript-legacy` and then we've made Webpack happy. In this
world, we can configure Webpack to run everthing in `app/javascript`
through the modern pipeline and everything in `app/javascript-legacy`
through it's legacy pipeline.

This kind of sucks but I think it's not as bad as having to have a rake
task handle this or actually copy the directory itself.

To handle the symlink, we are using `fs-extra` which adds a bunch of
extra functionality not found within the `fs` module.

When we produce the entries for Webpack, we are now returning 2
versions, `entries` which is all the entry points within
`app/javascript` and `legacyEntries` which is all the entry points
within `app/javascript-legacy`.

wannabefro committed:
Update default babel config to target modern browsers

We want to start producing 2 bundle of client JavaScript. 1 that targets
browsers that support `type=module` (modern browsers) and 1 that targets
browsers that don't.

This will lead to modern browsers receiving less bytes of JS over the
wire and lead to faster parsing and evaluation.

This commit updates the default babel config to target modern browsers
only (using the `esmodules: true` target).

One of the main changes here is that we are now using `useBuiltIns:
'usage` to handle polyfilling of JS language features.

This means that when babel encounters something that needs to be
polyfilled (it makes this decision based on the target) it includes the
necessary polyfill from `@babel/polyfills`.

For instance, if the browsers doesn't support `Promise` given

```
new Promise();
```

babel will rewrite it to look something like

```
import Promise from 'corejs/promise';

new Promise();
```

This means that we are:
a) only polyfilling what we use (as opposed to the whole JS language)
b) we don't have to worry about polyfilling the language (we still need
to worry about polyfilling browsers apis though)

Since we're targeting modern browsers, I've also removed the
`fast-async` plugin (which rewrites `async await` to use promises) since
all browsers that support `type="modules"` support `async await`.

I've also removed `object-rest-spread` for the same reasons.

To allow us to have multiple babel configs we are now also explicitly
stating which babel config to use within the loader.

We've set `modules` to false which instructs babel to not transpile
`import` statements to `require` statements since Webpack understands
`import` statements and gives it the capabilities to statically analyze
our code.

wannabefro committed:
Create es5 babel config

This is the same as our previous default config with the exception that
our target is now `esmodules: false` (i.e. browsers that don't support
`type=module`)

We have also introduced a new webpack loader that uses this config so
that we can apply it in a future commit to our legacy bundles.

wannabefro committed:
Create node babel config

This updates the targets to be `node` which should allow us to have a
more performant server side bundle. `node: 'current'` means that babel
will target the current version of node that is available on the
machine.

Since our version of node supports `async await` and
`object-rest-spread` I've removed those plugins from this config as
well.

wannabefro committed:
Rename default export from loaders

Since we are going to need to export multiple loader configs from this
file (to support the new node and es5 babel configs) this commit makes
it so that we are now exporting named exports from our loaders (and
updates the relevant places where we are using them).

wannabefro committed:
Introduce shared config for legacy bundles

We need a different base config for our Webpack config that's
responsible for producing our legacy bundles.

The main changes are:

* The entries need to be looking for our legacy entries (i.e.
`app/javascript-legacy`)
* Our resolved paths need to factor in `app/javascript-legacy`
* We need to use the babel config targeting ES5
* We need to suffix the outputs with `es5` to make it easy to tell which
bundle we are loading in the browser.

This commit adds theses changes.

wannabefro committed:
Development Legacy bundle

The only changes here are that we are using our `sharedLegacy` config
that we introduced in the previous commit and have changed the name of
the manifest to be `manifest-legacy.json` (as opposed to the default
`manifest.json`)

wannabefro committed:
Production legacy bundle

This modifies our production config to

1. Use shared-legacy
2. Rename the manifest to `manifest-legacy.json`
3. Removed `MiniCssExtractPlugin` (since we ignore sass and css in our
legacy bundles)

wannabefro committed:
Bump modern production config terser ecma target

Given that this targeting modern browsers, we can raise the ecma version
safely.

wannabefro committed:
3.0.0

sebjacobs commented on src/config.js:
```diff
@@ -3,39 +3,54 @@ const {
         const entryDir = dirname(path);
         const entryDirWithoutRootPath = entryDir.replace(rootPath, '');
-        results[`${entryDirWithoutRootPath}`] = resolve(path);
+        output[`${entryDirWithoutRootPath}`] = resolve(path);
```
As we spoke about earlier the fact that this is mutating either the `legacyEntries` or the `entries` objects defined  on :16/:17 could be confusing.

sebjacobs commented on src/hypernova.js:
```diff
@@ -46,17 +47,19 @@ const hypernovaConfig = {
-      },
-    })],
-  }});
+  Object.assign(hypernovaConfig, {
```
As we spoke about earlier it might clearer to split this change out since it might not be clear to someone that the `optimization` options haven't actually changed.

sebjacobs reviewed (approved):
Thanks for taking me through these changes and the changes to the main app!

Given that `src/config.js` is already a pretty confusing file I think it is worth spending a bit of time making the new changes introduced in this branch clearer.

It is a shame that the only thing which makes this possible is a big hack using symlinks (introduced in the first commit), but the fact that it works is very impressive :smile:!

wannabefro committed:
Update hypernova config to use node babel config

This introduces a new export to our loaders that is specific for node
and updates our hypernova config to use this new export.

wannabefro committed:
Make eslint happy with for hypernova optimisations
